bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookKeeperAdmin.java

```diff
@@ -810,47 +780,71 @@ public void openComplete(int newrc, final LedgerHandle newlh, Object newctx) {
+                        if (!dryrun) {
+                            ledgerFragmentsMcb.processResult(BKException.Code.NotEnoughBookiesException, null, null);
+                        } else {
+                            System.out.println("  Fragment [" + startEntryId + " - " + endEntryId + " ] : "
+                                               + BKException.getMessage(BKException.Code.NotEnoughBookiesException));
+                        }
+                        continue;
+                    }
+
+                    if (dryrun) {
+                        ArrayList<BookieSocketAddress> newEnsemble =
+                                replaceBookiesInEnsemble(ensemble, targetBookieAddresses);
+                        System.out.println("  Fragment [" + startEntryId + " - " + endEntryId + " ] : ");
+                        System.out.println("    old ensemble : " + formatEnsemble(ensemble, bookiesSrc, '*'));
+                        System.out.println("    new ensemble : " + formatEnsemble(newEnsemble, bookiesSrc, '*'));
                         continue;
                     }
 
                     if (LOG.isDebugEnabled()) {
                         LOG.debug("Replicating fragment from [" + startEntryId
                                   + "," + endEntryId + "] of ledger " + lh.getId()
-                                  + " to " + newBookie);
+                                  + " to " + targetBookieAddresses);
                     }
                     try {
                         LedgerFragmentReplicator.SingleFragmentCallback cb = new LedgerFragmentReplicator.SingleFragmentCallback(
-                                                                               ledgerFragmentsMcb, lh, startEntryId, bookieSrc, newBookie);
-                        ArrayList<BookieSocketAddress> currentEnsemble = lh.getLedgerMetadata().getEnsemble(
-                                startEntryId);
-                        int bookieIndex = -1;
-                        if (null != currentEnsemble) {
-                            for (int i = 0; i < currentEnsemble.size(); i++) {
-                                if (currentEnsemble.get(i).equals(bookieSrc)) {
-                                    bookieIndex = i;
-                                    break;
-                                }
-                            }
-                        }
+                                ledgerFragmentsMcb, lh, startEntryId, getReplacedBookiesMap(ensemble, targetBookieAddresses));
                         LedgerFragment ledgerFragment = new LedgerFragment(lh,
-                                startEntryId, endEntryId, bookieIndex);
-                        asyncRecoverLedgerFragment(lh, ledgerFragment, cb, newBookie);
+                                startEntryId, endEntryId, targetBookieAddresses.keySet());
+                        asyncRecoverLedgerFragment(lh, ledgerFragment, cb, Sets.newHashSet(targetBookieAddresses.values()));
```

Comment:
It looks to me that you are recovering the segments even on a dryrun.

Reply:
no. for each ledger, `dryrun` stops at printing messages at line 804.

Reply:
Ah, the continue doesn't pop out, and it doesn't help that dryrun is then used at line 823, which I guess is dead code. This could be clarified by using else rather continue, and maybe moving the contents of the else into a new method. This method is far too long as it is.

Reply:
sure, will move the else contents into a method.